From afaa29ccb5410bd1576e4412515a9403202e1c8c Mon Sep 17 00:00:00 2001
From: Phong Hoang <phong.hoang.wz@renesas.com>
Date: Thu, 15 Apr 2021 11:23:56 +0700
Subject: [PATCH 11/40] remoteproc: use dynamic allocation if bad carveout rsc
 configuration

In case bad carveout rsc configuration and coprocessor don't build
pa to da itself, should try to use dynamic allocation for avoiding
unexpected behaviors.

Fixes: 60f849a5c15 (remoteproc: "fix rproc_alloc_carveout() for rproc with iommu domain")

Signed-off-by: Phong Hoang <phong.hoang.wz@renesas.com>
---
 drivers/remoteproc/remoteproc_core.c | 13 ++++++++++---
 include/linux/remoteproc.h           |  2 ++
 2 files changed, 12 insertions(+), 3 deletions(-)

diff --git a/drivers/remoteproc/remoteproc_core.c b/drivers/remoteproc/remoteproc_core.c
index 564bb5ccb64e..8fd07892576c 100644
--- a/drivers/remoteproc/remoteproc_core.c
+++ b/drivers/remoteproc/remoteproc_core.c
@@ -813,9 +813,15 @@ static int rproc_alloc_carveout(struct rproc *rproc,
 		 * Don't stop rproc_start sequence as coprocessor may
 		 * build pa to da translation on its side.
 		 */
-		if (mem->da != (u32)dma)
-			dev_warn(dev->parent,
-				 "Allocated carveout doesn't fit device address request\n");
+		if (mem->da != (u32)dma) {
+			if (rproc->has_pa_to_da) {
+				dev_warn(dev->parent,
+					 "Allocated carveout doesn't fit device address request\n");
+			} else {
+				dev_dbg(dev, "Bad carveout rsc configuration. Try to use dynamic alloc.\n");
+				goto dynamic_alloc;
+			}
+		}
 	}
 
 	/*
@@ -865,6 +871,7 @@ static int rproc_alloc_carveout(struct rproc *rproc,
 	}
 
 	if (mem->da == FW_RSC_ADDR_ANY) {
+dynamic_alloc:
 		/* Update device address as undefined by requester */
 		if ((u64)dma & HIGH_BITS_MASK)
 			dev_warn(dev, "DMA address cast in 32bit to fit resource table format\n");
diff --git a/include/linux/remoteproc.h b/include/linux/remoteproc.h
index 185d435b7155..48bb677f1481 100644
--- a/include/linux/remoteproc.h
+++ b/include/linux/remoteproc.h
@@ -510,6 +510,7 @@ struct rproc_dump_segment {
  * @cached_table: copy of the resource table
  * @table_sz: size of @cached_table
  * @has_iommu: flag to indicate if remote processor is behind an MMU
+ * @has_pa_to_da: flag to indicate if remote processor translates pa to da itself
  * @auto_boot: flag to indicate if remote processor should be auto-started
  * @autonomous: true if an external entity has booted the remote processor
  * @skip_fw_load: remote processor has been preloaded before start sequence
@@ -548,6 +549,7 @@ struct rproc {
 	struct resource_table *cached_table;
 	size_t table_sz;
 	bool has_iommu;
+	bool has_pa_to_da;
 	bool auto_boot;
 	bool autonomous;
 	bool skip_fw_load;
-- 
2.25.1

